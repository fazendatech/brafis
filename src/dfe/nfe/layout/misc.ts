import { z } from "zod";
import { ufCodeMap } from "@/ufCode";
import type { UF, UFCode } from "@/ufCode/types";
import type { NfeLayout } from ".";

// NOTE: Hack pro zod aceitar tipos customizados (https://stackoverflow.com/a/73825370)
const ufCodeList = Object.values(ufCodeMap);
const ufCodes: [UFCode, ...UFCode[]] = [ufCodeList[0], ...ufCodeList];
export const zUfCode = () => z.enum(ufCodes);

const ufList = Object.keys(ufCodeMap);
const ufs: [UF, ...UF[]] = [ufList[0], ...ufList];
export const zUf = () => z.enum(ufs);

// NOTE: NFe válida usada para testes
export const NFE_TEST_DATA: NfeLayout = {
  NFe: {
    "@_xmlns": "http://www.portalfiscal.inf.br/nfe",
    infNFe: {
      "@_versao": "4.00",
      "@_Id": "NFe12345678901234567890123456789012345678901235",
      ide: {
        cUF: "35",
        cNF: "00012345",
        natOp: "Venda",
        mod: "55",
        serie: "001",
        nNF: "123456",
        dhEmi: "2024-12-06T08:00:00Z",
        tpNF: "1",
        idDest: "1",
        cMunFG: "3550308",
        tpImp: "1",
        tpEmis: "1",
        cDV: "5",
        tpAmb: "1",
        finNFe: "1",
        indFinal: "1",
        indPres: "1",
        procEmi: "0",
        verProc: "1.0.0",
        NFref: [
          {
            refNFe: "12345678901234567890123456789012345678901234",
          },
          {
            refNF: {
              cUF: "35",
              AAMM: "2412",
              CNPJ: "12345678000195",
              mod: "01",
              serie: "001",
              nNF: "123456",
            },
          },
          {
            refNFP: {
              cUF: "35",
              AAMM: "2412",
              CPF: "11155599900",
              IE: "1234567890",
              mod: "04",
              serie: "001",
              nNF: "123456",
            },
          },
          {
            refCTe: "12345678901234567890123456789012345678901234",
          },
          {
            refECF: {
              mod: "2B",
              nECF: "123",
              nCOO: "123456",
            },
          },
        ],
      },
      emit: {
        CNPJ: "12345678000195",
        xNome: "Empresa Emitente LTDA",
        xFant: "Fantasia Emitente",
        enderEmit: {
          xLgr: "Rua Exemplo",
          nro: "123",
          xBairro: "Bairro Exemplo",
          cMun: "3550308",
          xMun: "São Paulo",
          UF: "SP",
          CEP: "01001000",
        },
        IE: "1234567890",
        CRT: "3",
      },
      avulsa: {
        CNPJ: "12345678000195",
        xOrgao: "Secretaria da Fazenda",
        matr: "123456789",
        xAgente: "Agente da Fazenda",
        fone: "11999999999",
        UF: "SP",
        nDAR: "123456789012345678901234567890123456789012345678901234567890",
        dEmi: "2024-12-06",
        vDAR: "500.00",
        repEmi: "0",
        dPag: "2024-12-06",
      },
      dest: {
        CPF: "11155599900",
        xNome: "Destinatário Exemplo",
        enderDest: {
          xLgr: "Avenida Exemplo",
          nro: "456",
          xBairro: "Outro Bairro",
          cMun: "3304557",
          xMun: "Rio de Janeiro",
          UF: "RJ",
          CEP: "20040030",
        },
        indIEDest: "9",
      },
      retirada: {
        CNPJ: "12345678000195",
        xLgr: "Rua Exemplo",
        nro: "123",
        xBairro: "Bairro Exemplo",
        cMun: "3550308",
        xMun: "São Paulo",
        UF: "SP",
      },
      entrega: {
        CNPJ: "12345678000195",
        xLgr: "Rua Exemplo",
        nro: "123",
        xBairro: "Bairro Exemplo",
        cMun: "3550308",
        xMun: "São Paulo",
        UF: "SP",
      },
      autXML: [
        {
          CNPJ: "12345678000195",
        },
      ],
      det: [
        {
          "@_nItem": "1",
          prod: {
            cProd: "001",
            cEAN: "SEM GTIN",
            xProd: "Produto Exemplo",
            NCM: "12345678",
            CFOP: "5102",
            uCom: "UN",
            qCom: "10.0000",
            vUnCom: "50.0000000000",
            vProd: "500.00",
            uTrib: "UN",
            qTrib: "10.0000",
            vUnTrib: "50.0000000000",
            indTot: "1",
            DI: [
              {
                nDI: "123456",
                dDI: "2024-12-06",
                xLocDesemb: "Local Desembaraço",
                UFDesemb: "SP",
                dDesemb: "2024-12-06",
                cExportador: "123456789",
                adi: [
                  {
                    nAdicao: "1",
                    nSeqAdic: "1",
                    cFabricante: "123456789",
                    vDescDI: "0.00",
                    nDraw: "12345678901",
                  },
                ],
                tpViaTransp: "7",
                tpIntermedio: "1",
              },
            ],
            detExport: [
              {
                nDraw: "12345678901",
                exportInd: {
                  nRE: "123456789012",
                  chNFe: "12345678901212345678901234567890123456789012",
                  qExport: "10.000",
                },
              },
            ],
            rastro: [
              {
                nLote: "1",
                qLote: "10.000",
                dFab: "2024-12-06",
                dVal: "2024-12-06",
                cAgreg: "123456789",
              },
            ],
          },
          imposto: {
            ICMS: {
              ICMS00: {
                orig: "0",
                CST: "00",
                modBC: "3",
                vBC: "500.00",
                pICMS: "18.00",
                vICMS: "90.00",
              },
            },
          },
        },
      ],
      total: {
        ICMSTot: {
          vBC: "500.00",
          vICMS: "90.00",
          vICMSDeson: "0.00",
          vFCP: "0.00",
          vBCST: "0.00",
          vST: "0.00",
          vFCPST: "0.00",
          vFCPSTRet: "0.00",
          vProd: "500.00",
          vFrete: "0.00",
          vSeg: "0.00",
          vDesc: "0.00",
          vII: "0.00",
          vIPI: "0.00",
          vIPIDevol: "0.00",
          vPIS: "0.00",
          vCOFINS: "0.00",
          vOutro: "0.00",
          vNF: "500.00",
        },
        ISSQNtot: {
          vServ: "0.00",
          vBC: "0.00",
          vISS: "0.00",
          vPIS: "0.00",
          vCOFINS: "0.00",
          dCompet: "2024-12-06",
          vDeducao: "0.00",
          vOutro: "0.00",
          vDescIncond: "0.00",
          vDescCond: "0.00",
          vISSRet: "0.00",
          cRegTrib: "5",
        },
        retTrib: {
          vRetPIS: "0.00",
          vRetCOFINS: "0.00",
          vRetCSLL: "0.00",
          vBCIRRF: "0.00",
          vIRRF: "0.00",
          vBCRetPrev: "0.00",
          vRetPrev: "0.00",
        },
      },
      transp: {
        modFrete: "1",
      },
      pag: {
        detPag: [
          {
            tPag: "01",
            vPag: "500.00",
          },
        ],
      },
      cobr: {
        dup: [
          {
            vDup: "500.00",
          },
        ],
      },
      infIntermed: {
        CNPJ: "00023456000177",
        idCadIntTran:
          "123456789012345678901234567890123456789012345678901234567890",
      },
      infAdic: {},
      infRespTec: {
        CNPJ: "12345678000195",
        xContato: "Responsável Técnico",
        email: "responsavel@empresa.com",
        fone: "11999999999",
      },
    },
  },
};
