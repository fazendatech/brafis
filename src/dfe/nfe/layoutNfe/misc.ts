import { z } from "zod";
import { ufCodeMap } from "@/ufCode";
import type { UF, UFCode } from "@/ufCode/types";

// NOTE: Hack pro zod aceitar tipos customizados (https://stackoverflow.com/a/73825370)
const ufCodeList = Object.values(ufCodeMap);
const ufCodes: [UFCode, ...UFCode[]] = [ufCodeList[0], ...ufCodeList];
export const zUfCode = () => z.enum(ufCodes);

const ufList = Object.keys(ufCodeMap);
const ufs: [UF, ...UF[]] = [ufList[0], ...ufList];
export const zUf = () => z.enum(ufs);
