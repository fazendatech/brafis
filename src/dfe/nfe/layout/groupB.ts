import { zCustom } from "@/utils/zCustom";
import { z } from "zod";
import { schemaNfeNfRef } from "./groupBa";
import { zUfCode } from "./misc";

const invalidCnfs = new Set([
  "00000000",
  "11111111",
  "22222222",
  "33333333",
  "44444444",
  "55555555",
  "66666666",
  "77777777",
  "88888888",
  "99999999",
  "12345678",
  "23456789",
  "34567890",
  "45678901",
  "56789012",
  "67890123",
  "78901234",
  "89012345",
  "90123456",
  "01234567",
]);

const schemaNfeIde = z
  .object({
    cUF: zUfCode().describe("B02"),
    cNF: zCustom
      .numeric()
      .length(8)
      .refine((cNF) => !invalidCnfs.has(cNF), {
        message: "Código numérico em formato inválido.",
      })
      .describe("B03"),
    natOp: zCustom.length(1, 60).describe("B04"),
    mod: z.enum(["55", "65"]).describe("B06"),
    serie: zCustom.numeric().length(3).describe("B07"),
    nNF: zCustom.numeric().min(1).max(9).describe("B08"),
    dhEmi: zCustom.date().describe("B09"),
    dhSaiEnt: zCustom.date().optional().describe("B10"),
    tpNF: z.enum(["0", "1"]).describe("B11"),
    idDest: z.enum(["1", "2", "3"]).describe("B11a"),
    cMunFG: zCustom.numeric().length(7).describe("B12"),
    tpImp: z.enum(["0", "1", "2", "3", "4", "5"]).describe("B21"),
    tpEmis: z.enum(["1", "2", "3", "4", "5", "6", "7", "9"]).describe("B22"),
    cDV: zCustom.numeric().length(1).describe("B23"),
    tpAmb: z.enum(["1", "2"]).describe("B24"),
    finNFe: z.enum(["1", "2", "3", "4"]).describe("B25"),
    indFinal: z.enum(["0", "1"]).describe("B25a"),
    indPres: z.enum(["0", "1", "2", "3", "4", "5", "9"]).describe("B25b"),
    indIntermed: z.enum(["0", "1"]).optional().describe("B25c"),
    procEmi: z.enum(["0", "1", "2", "3"]).describe("B26"),
    verProc: zCustom.length(1, 20).describe("B27"),
    dhCont: zCustom.date().optional().describe("B28"),
    xJust: zCustom.length(15, 256).optional().describe("B29"),
    NFref: z.array(schemaNfeNfRef).max(500).optional(),
  })
  .refine(({ mod, dhSaiEnt }) => (mod === "65" ? !dhSaiEnt : true), {
    message:
      "Data e hora de Saída ou da Entrada da Mercadoria/Produto não deve ser informada para NFC-e.",
  })
  .refine(
    ({ tpNF, dhSaiEnt, dhEmi }) => {
      if (!dhSaiEnt || tpNF !== "1") {
        return true;
      }
      return new Date(dhSaiEnt) >= new Date(dhEmi);
    },
    {
      message:
        "Data e hora de Saída ou da Entrada da Mercadoria/Produto deve ser maior ou igual a Data e hora de Emissão.",
    },
  )
  .refine(({ cNF, nNF }) => Number(cNF) !== Number(nNF), {
    message: "O código numérico da NF-e não pode ser igual ao número da NF-e.",
  })
  .describe("ide:B01");

type NfeIde = z.infer<typeof schemaNfeIde>;

/**
 * @description Grupo B. Identificação da Nota Fiscal eletrônica
 */
export { schemaNfeIde, type NfeIde };
