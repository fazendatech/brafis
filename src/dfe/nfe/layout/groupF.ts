import { zCustom } from "@/utils/zCustom";
import { z } from "zod";
import { zUf } from "./misc";

const schemaNfeRetirada = z
  .object({
    CNPJ: zCustom.cnpj().optional().describe("F02"),
    CPF: zCustom.cpf().optional().describe("F02a"),
    xNome: zCustom.length(2, 60).optional().describe("F02b"),
    xLgr: zCustom.length(2, 60).describe("F03"),
    nro: zCustom.length(1, 60).describe("F04"),
    xCpl: zCustom.length(1, 60).optional().describe("F05"),
    xBairro: zCustom.length(2, 60).describe("F06"),
    cMun: zCustom.numeric().length(7).describe("F07"),
    xMun: zCustom.length(2, 60).describe("F08"),
    UF: zUf().or(z.literal("EX")).describe("F09"),
    CEP: zCustom.numeric().length(8).optional().describe("F10"),
    cPais: zCustom.numeric().length(4).optional().describe("F11"),
    xPais: zCustom.length(2, 60).optional().describe("F12"),
    fone: zCustom.phone().optional().describe("F13"),
    email: zCustom.length(1, 60).email().optional().describe("F14"),
    IE: zCustom.ie().optional().describe("F15"),
  })
  .refine(({ CNPJ, CPF }) => zCustom.utils.hasOnlyOne(CNPJ, CPF), {
    message: "Deve ser informado apenas um dos campos: CNPJ ou CPF.",
  })
  .refine(
    ({ cMun, xMun, UF }) => {
      if (cMun === "9999999" && xMun === "EXTERIOR" && UF === "EX") {
        return true;
      }
      if (cMun !== "9999999" && xMun !== "EXTERIOR" && UF !== "EX") {
        return true;
      }
      return false;
    },
    {
      message:
        "Se a operação for no exterior, informar cMun = '9999999', xMun = 'EXTERIOR' e UF = 'EX'.",
    },
  )
  .describe("retirada:F01");

/**
 * @description Grupo F. Identificação do Local de Retirada
 */
type NfeRetirada = z.infer<typeof schemaNfeRetirada>;

export { schemaNfeRetirada, type NfeRetirada };
