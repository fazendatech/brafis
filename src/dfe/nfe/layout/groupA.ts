import { isValidAccessCode } from "@/utils/validators/isValidAccessCode";
import { z } from "zod";
import { schemaNfeIde } from "./groupB";
import { schemaNfeEmit } from "./groupC";
import { schemaNfeAvulsa } from "./groupD";
import { schemaNfeDest } from "./groupE";
import { schemaNfeRetirada } from "./groupF";
import { schemaNfeEntrega } from "./groupG";
import { schemaNfeAutXml } from "./groupGa";
import { schemaNfeDet } from "./groupH";
import { schemaNfeTotal } from "./groupW";
import { schemaNfeTransp } from "./groupX";
import { schemaNfeCobr } from "./groupY";
import { schemaNfePag } from "./groupYa";
import { schemaNfeInfIntermed } from "./groupYb";
import { schemaNfeInfAdic } from "./groupZ";
import { schemaNfeInfRespTec } from "./groupZd";

const schemaNfeInfNfe = z
  .object({
    "@_versao": z.literal("4.00").describe("A02"),
    "@_Id": z
      .string()
      .startsWith("NFe")
      .length(47)
      .refine((value) => isValidAccessCode(value.slice(3), { strict: true }), {
        message: "O Id da nota está em formato inválido.",
      })
      .describe("A03"),
    ide: schemaNfeIde,
    emit: schemaNfeEmit,
    avulsa: schemaNfeAvulsa.optional(),
    dest: schemaNfeDest,
    retirada: schemaNfeRetirada.optional(),
    entrega: schemaNfeEntrega.optional(),
    autXML: z.array(schemaNfeAutXml).max(10).optional(),
    det: z
      .array(schemaNfeDet)
      .min(1)
      .max(990)
      .refine(
        (array) =>
          array.every(({ "@_nItem": nItem }, i) => Number(nItem) === i + 1),
        { message: "nItem deve ser sequencial a partir de 1" },
      ),
    total: schemaNfeTotal,
    transp: schemaNfeTransp,
    cobr: schemaNfeCobr.optional(),
    pag: schemaNfePag,
    infIntermed: schemaNfeInfIntermed.optional(),
    infAdic: schemaNfeInfAdic.optional(),
    infRespTec: schemaNfeInfRespTec.optional(),
  })
  .refine(({ ide, dest }) => (ide.mod === "55" ? !!dest.xNome : true), {
    message: "dest.xNome é obrigatório para a NF-e (modelo 55).",
  })
  .refine(
    ({ ide, dest }) =>
      ide.mod === "65" ? dest.indIEDest === "9" && !dest.IE : true,
    {
      message:
        "No caso de NFC-e (modelo 65) informar dest.indIEDest=9 e não informar a tag dest.IE",
    },
  )
  .refine(({ ide, "@_Id": id }) => ide.cDV === id.slice(-1), {
    message:
      "O dígito verificador do Id da nota deve ser igual ao campo cDV da tag ide.",
  })
  .describe("infNFe:A01");

/**
 * @description Grupo A. Dados da Nota Fiscal eletrônica
 */
type NfeInfNfe = z.infer<typeof schemaNfeInfNfe>;

export { schemaNfeInfNfe, type NfeInfNfe };
