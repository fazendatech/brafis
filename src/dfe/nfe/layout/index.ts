import { z } from "zod";
import { schemaNfeInfNfe } from "./groupA";

const schemaNfe = z
  .object({
    NFe: z.object({
      "@_xmlns": z.literal("http://www.portalfiscal.inf.br/nfe"),
      infNFe: schemaNfeInfNfe,
    }),
  })
  .describe("NFe (TAG Raiz)");

/**
 * @description Tipo que infere a estrutura da NFe.
 */
export type NfeLayout = z.infer<typeof schemaNfe>;

/**
 * @description Realiza a validação da NFe.
 *
 * @param {NfeLayout} nfe - Objeto NFe.
 *
 * @returns {NfeLayout} O objeto da NFe validado.
 * @throws {Zod.ZodError} Se a NFe não for válida.
 */
export function parseNfe(nfe: NfeLayout): NfeLayout {
  return schemaNfe.parse(nfe);
}
