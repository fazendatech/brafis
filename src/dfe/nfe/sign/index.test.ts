import { describe, expect, test } from "bun:test";
import { CertificateP12 } from "@/certificate";
import { signNfe } from ".";
import { NFE_TEST_DATA } from "@/dfe/nfe/layout/misc";
import { makeParser, makeBuilder } from "@/utils/xml";

describe("sign", () => {
  describe("signNfe", async () => {
    const certificate = await CertificateP12.fromFilepath({
      filepath: "misc/sample-certificates/cert.pfx",
      password: "senha",
    });

    test("Signs NFe XML correctly", () => {
      expect(signNfe(NFE_TEST_DATA, certificate)).toMatchSnapshot();
    });

    test("Parses and rebuilds signed NFe correctly", () => {
      const signedNfe = signNfe(NFE_TEST_DATA, certificate);
      const signedNfeObject = makeParser().parse(signedNfe);

      expect(makeBuilder().build(signedNfeObject)).toEqual(signedNfe);
    });
  });
});
