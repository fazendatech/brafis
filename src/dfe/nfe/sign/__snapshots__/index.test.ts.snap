// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`sign signNfe Returns true when generate and sign the XML correctly 1`] = `"<NFe xmlns="http://www.portalfiscal.inf.br/nfe"><infNFe versao="4.00" Id="NFe12345678901234567890123456789012345678901235"><ide><cUF>35</cUF><cNF>00012345</cNF><natOp>Venda</natOp><mod>55</mod><serie>001</serie><nNF>123456</nNF><dhEmi>2024-12-06T08:00:00Z</dhEmi><tpNF>1</tpNF><idDest>1</idDest><cMunFG>3550308</cMunFG><tpImp>1</tpImp><tpEmis>1</tpEmis><cDV>5</cDV><tpAmb>1</tpAmb><finNFe>1</finNFe><indFinal>1</indFinal><indPres>1</indPres><procEmi>0</procEmi><verProc>1.0.0</verProc><NFref><refNFe>12345678901234567890123456789012345678901234</refNFe></NFref><NFref><refNF><cUF>35</cUF><AAMM>2412</AAMM><CNPJ>12345678000195</CNPJ><mod>01</mod><serie>001</serie><nNF>123456</nNF></refNF></NFref><NFref><refNFP><cUF>35</cUF><AAMM>2412</AAMM><CPF>11155599900</CPF><IE>1234567890</IE><mod>04</mod><serie>001</serie><nNF>123456</nNF></refNFP></NFref><NFref><refCTe>12345678901234567890123456789012345678901234</refCTe></NFref><NFref><refECF><mod>2B</mod><nECF>123</nECF><nCOO>123456</nCOO></refECF></NFref></ide><emit><CNPJ>12345678000195</CNPJ><xNome>Empresa Emitente LTDA</xNome><xFant>Fantasia Emitente</xFant><enderEmit><xLgr>Rua Exemplo</xLgr><nro>123</nro><xBairro>Bairro Exemplo</xBairro><cMun>3550308</cMun><xMun>São Paulo</xMun><UF>SP</UF><CEP>01001000</CEP></enderEmit><IE>1234567890</IE><CRT>3</CRT></emit><avulsa><CNPJ>12345678000195</CNPJ><xOrgao>Secretaria da Fazenda</xOrgao><matr>123456789</matr><xAgente>Agente da Fazenda</xAgente><fone>11999999999</fone><UF>SP</UF><nDAR>123456789012345678901234567890123456789012345678901234567890</nDAR><dEmi>2024-12-06</dEmi><vDAR>500.00</vDAR><repEmi>0</repEmi><dPag>2024-12-06</dPag></avulsa><dest><CPF>11155599900</CPF><xNome>Destinatário Exemplo</xNome><enderDest><xLgr>Avenida Exemplo</xLgr><nro>456</nro><xBairro>Outro Bairro</xBairro><cMun>3304557</cMun><xMun>Rio de Janeiro</xMun><UF>RJ</UF><CEP>20040030</CEP></enderDest><indIEDest>9</indIEDest></dest><retirada><CNPJ>12345678000195</CNPJ><xLgr>Rua Exemplo</xLgr><nro>123</nro><xBairro>Bairro Exemplo</xBairro><cMun>3550308</cMun><xMun>São Paulo</xMun><UF>SP</UF></retirada><entrega><CNPJ>12345678000195</CNPJ><xLgr>Rua Exemplo</xLgr><nro>123</nro><xBairro>Bairro Exemplo</xBairro><cMun>3550308</cMun><xMun>São Paulo</xMun><UF>SP</UF></entrega><autXML><CNPJ>12345678000195</CNPJ></autXML><det nItem="1"><prod><cProd>001</cProd><cEAN>SEM GTIN</cEAN><xProd>Produto Exemplo</xProd><NCM>12345678</NCM><CFOP>5102</CFOP><uCom>UN</uCom><qCom>10.0000</qCom><vUnCom>50.0000000000</vUnCom><vProd>500.00</vProd><uTrib>UN</uTrib><qTrib>10.0000</qTrib><vUnTrib>50.0000000000</vUnTrib><indTot>1</indTot><DI><nDI>123456</nDI><dDI>2024-12-06</dDI><xLocDesemb>Local Desembaraço</xLocDesemb><UFDesemb>SP</UFDesemb><dDesemb>2024-12-06</dDesemb><cExportador>123456789</cExportador><adi><nAdicao>1</nAdicao><nSeqAdic>1</nSeqAdic><cFabricante>123456789</cFabricante><vDescDI>0.00</vDescDI><nDraw>12345678901</nDraw></adi><tpViaTransp>7</tpViaTransp><tpIntermedio>1</tpIntermedio></DI><detExport><nDraw>12345678901</nDraw><exportInd><nRE>123456789012</nRE><chNFe>12345678901212345678901234567890123456789012</chNFe><qExport>10.000</qExport></exportInd></detExport><rastro><nLote>1</nLote><qLote>10.000</qLote><dFab>2024-12-06</dFab><dVal>2024-12-06</dVal><cAgreg>123456789</cAgreg></rastro></prod><imposto><ICMS><ICMS00><orig>0</orig><CST>00</CST><modBC>3</modBC><vBC>500.00</vBC><pICMS>18.00</pICMS><vICMS>90.00</vICMS></ICMS00></ICMS></imposto></det><total><ICMSTot><vBC>500.00</vBC><vICMS>90.00</vICMS><vICMSDeson>0.00</vICMSDeson><vFCP>0.00</vFCP><vBCST>0.00</vBCST><vST>0.00</vST><vFCPST>0.00</vFCPST><vFCPSTRet>0.00</vFCPSTRet><vProd>500.00</vProd><vFrete>0.00</vFrete><vSeg>0.00</vSeg><vDesc>0.00</vDesc><vII>0.00</vII><vIPI>0.00</vIPI><vIPIDevol>0.00</vIPIDevol><vPIS>0.00</vPIS><vCOFINS>0.00</vCOFINS><vOutro>0.00</vOutro><vNF>500.00</vNF></ICMSTot><ISSQNtot><vServ>0.00</vServ><vBC>0.00</vBC><vISS>0.00</vISS><vPIS>0.00</vPIS><vCOFINS>0.00</vCOFINS><dCompet>2024-12-06</dCompet><vDeducao>0.00</vDeducao><vOutro>0.00</vOutro><vDescIncond>0.00</vDescIncond><vDescCond>0.00</vDescCond><vISSRet>0.00</vISSRet><cRegTrib>5</cRegTrib></ISSQNtot><retTrib><vRetPIS>0.00</vRetPIS><vRetCOFINS>0.00</vRetCOFINS><vRetCSLL>0.00</vRetCSLL><vBCIRRF>0.00</vBCIRRF><vIRRF>0.00</vIRRF><vBCRetPrev>0.00</vBCRetPrev><vRetPrev>0.00</vRetPrev></retTrib></total><transp><modFrete>1</modFrete></transp><pag><detPag><tPag>01</tPag><vPag>500.00</vPag></detPag></pag><cobr><dup><vDup>500.00</vDup></dup></cobr><infIntermed><CNPJ>00023456000177</CNPJ><idCadIntTran>123456789012345678901234567890123456789012345678901234567890</idCadIntTran></infIntermed><infAdic/><infRespTec><CNPJ>12345678000195</CNPJ><xContato>Responsável Técnico</xContato><email>responsavel@empresa.com</email><fone>11999999999</fone></infRespTec></infNFe><Signature xmlns="http://www.w3.org/2000/09/xmldsig#"><SignedInfo><CanonicalizationMethod Algorithm="http://www.w3.org/TR/2001/REC-xml-c14n-20010315"/><SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/><Reference URI="#NFe12345678901234567890123456789012345678901235"><Transforms><Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/><Transform Algorithm="http://www.w3.org/TR/2001/REC-xml-c14n-20010315"/></Transforms><DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/><DigestValue>m2wNpmSlCY789bgA9oeRFHxrB4g=</DigestValue></Reference></SignedInfo><SignatureValue>WaTrwwmv6jP4NKFuV1JeekxSLfNJQRLqVxI8LtLHtTS274mm6aKlQ7W5sS/NjgYz9ZiWAHNxQOqT3HLkSVEYaevz0MMmrM5oeRKu5PJ4XcgzJF2V9QIyKi7wYhjkyy6tQt6a23qq4Uz5NHD11WlWp/8uQPJgtbeFhhGJ1WDuAxBGDFIiLfcztNWHOE+WApER8RTQ+Y+MVHLhgQLf0a8u0QFxSPny5LXrd6ZgZA0+hQiq4CcGoilPh8Wvm0mlZRIG0/6UHN+0UMPYZvJ56zYVEU3IQYJcKZshO+u9xMXExoLSHh4JXaCAtOdr+vdmnEAgXl6qNeo2foh1TIc3oKzipg==</SignatureValue><KeyInfo><X509Data><X509Certificate>MIIHsTCCBZmgAwIBAgIITIXNEVOBc9YwDQYJKoZIhvcNAQELBQAwdDELMAkGA1UEBhMCQlIxEzARBgNVBAoTCklDUC1CcmFzaWwxNjA0BgNVBAsTLVNlY3JldGFyaWEgZGEgUmVjZWl0YSBGZWRlcmFsIGRvIEJyYXNpbCAtIFJGQjEYMBYGA1UEAxMPQUMgVkFMSUQgUkZCIHY1MB4XDTIzMTIwNDExMzQxOFoXDTI0MTIwMzExMzQxOFowgfMxCzAJBgNVBAYTAkJSMRMwEQYDVQQKEwpJQ1AtQnJhc2lsMTYwNAYDVQQLEy1TZWNyZXRhcmlhIGRhIFJlY2VpdGEgRmVkZXJhbCBkbyBCcmFzaWwgLSBSRkIxFTATBgNVBAsTDFJGQiBlLUNQRiBBMTEYMBYGA1UECxMPQUMgVkFMSUQgUkZCIFY1MQ4wDAYDVQQLEwVBUiBKUjEZMBcGA1UECxMQVmlkZW9jb25mZXJlbmNpYTEXMBUGA1UECxMOMTk1NzY1NjcwMDAxMzkxIjAgBgNVBAMTGUZFTElQUEUgTEVOWkk6MDMyMTY4MDE5NzcwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDELjwh4k19/dqdc7m0O/R+j67UUuyvggzeItjZBTudOnQ5uYwf2x08FnswFF+/jC1FDitsTYb1eFgifLS7wFSlIdk7qQykZY13gmn5rRBrTixAmK9FKYAHm8ezPAyE57stca68mZVT9PbW97Wi8dHLCJ6KFEsU/uALNulUILY8xCe1Ip8Rua8yJ3NjnwfUl27LU50uJsc+1Bonh+sylRFWiwtf/y/AVpr+zoWwBc9Tr5ElWf3lmSEOSNaiEKla5bMcEOZd57ncgJgubttx0KhNm4/Aj3wxXzXnimto1s2SUp578A5QUAcwfnvs9luf7sGVKdMw6mijx0eYa82xY1gTAgMBAAGjggLFMIICwTCBnAYIKwYBBQUHAQEEgY8wgYwwVQYIKwYBBQUHMAKGSWh0dHA6Ly9pY3AtYnJhc2lsLnZhbGlkY2VydGlmaWNhZG9yYS5jb20uYnIvYWMtdmFsaWRyZmIvYWMtdmFsaWRyZmJ2NS5wN2IwMwYIKwYBBQUHMAGGJ2h0dHA6Ly9vY3NwdjUudmFsaWRjZXJ0aWZpY2Fkb3JhLmNvbS5icjAJBgNVHRMEAjAAMB8GA1UdIwQYMBaAFFPLpeR1UJlALL5bFUXJvsswqonFMHAGA1UdIARpMGcwZQYGYEwBAgElMFswWQYIKwYBBQUHAgEWTWh0dHA6Ly9pY3AtYnJhc2lsLnZhbGlkY2VydGlmaWNhZG9yYS5jb20uYnIvYWMtdmFsaWRyZmIvZHBjLWFjLXZhbGlkcmZiVjUucGRmMIG2BgNVHR8Ega4wgaswU6BRoE+GTWh0dHA6Ly9pY3AtYnJhc2lsLnZhbGlkY2VydGlmaWNhZG9yYS5jb20uYnIvYWMtdmFsaWRyZmIvbGNyLWFjLXZhbGlkcmZidjUuY3JsMFSgUqBQhk5odHRwOi8vaWNwLWJyYXNpbDIudmFsaWRjZXJ0aWZpY2Fkb3JhLmNvbS5ici9hYy12YWxpZHJmYi9sY3ItYWMtdmFsaWRyZmJ2NS5jcmwwDgYDVR0PAQH/BAQDAgXgMB0GA1UdJQQWMBQGCCsGAQUFBwMCBggrBgEFBQcDBDCBmQYDVR0RBIGRMIGOgRlzYW50YWFuYWZhemVuZGFAZ21haWwuY29toDgGBWBMAQMBoC8ELTA5MDgxOTc4MDMyMTY4MDE5NzcwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMKAXBgVgTAEDBqAOBAwwMzIxNjgwMTkwMDGgHgYFYEwBAwWgFQQTMDAwMDAwMDAwMDAwMDAwMDAwMDANBgkqhkiG9w0BAQsFAAOCAgEAWFdTesOkJnVpv2KeAO8OuwCRSmgBQ5XKzJYaa+AuIXPYVV3JpGOxzH4vM7UfWvureke+zWJW/CuY4oWiAJj4ComrsvRShbS+ybvNJ1i3PaTiEiJ6LoUv2wDHz12NWp3xHhNyReAUqeLByp97ICvxGDNKPBXSxoB0LVKscSaWWQa6tUU5Rp+OtUu0/GMm2YM3RYqMphvb/rG+cddmhQKTGcYn6QuJ1IO0X5ku+gbKmRM5E8d/3ZFogh+riWjRrbMpGE20IE7nFuzVxoNxPMAq6LTngaf1CfBhb/cOPSALBpZ0tHktxcK5Z3d8blM+0Y7J3R6Tg09cOti4REd1CgqdSq7u44+P0/PuZAdVKpWowhlzWtxCk+PshtNEvtCfSDl7tumaMUmLpqMQET7TXKgh+zQgogN1pS/lUZSbFtKcxqaheSDrAqKc7642Mj7A4p51o0ktJIo7DgP0zwKE2jLycN5ijgQon/YnmQyAiUDVUajDgtwjB/+q/9ij8dRKUAjTgv4GIwFvHtfPaaaxWDmEjfD33QVhfSE2rebOy9Sz/GYQ9aIvf4sweyBzfXFvfzAOah/moBbZKeSIvT7cZywk/OT1IYl+E3woqlVkksiVHmGXiwcIrpHrO21WB12dDY4nWa4TdzJwLSklguXGkHNmHDsb1s0FRDvSxKOU1OMhJ+4=</X509Certificate></X509Data></KeyInfo></Signature></NFe>"`;
